src/parser/regexp.bnf

```diff
@@ -431,19 +443,68 @@ function GroupRefOrDecChar(text, textLoc) {
 /**
  * Unicode names.
  */
-const uRe = /^\\u[0-9a-fA-F]{4}/;
-const ucpRe = /^\\u\{[0-9a-fA-F]{1,}\}/;
+const uReStart = /^\\u[0-9a-fA-F]{4}/; // only matches start of string
+const ucpReStart = /^\\u\{[0-9a-fA-F]{1,}\}/; // only matches start of string
+const ucpReAnywhere = /\\u\{[0-9a-fA-F]{1,}\}/; // matches anywhere in string
 
 /**
  * Validates Unicode group name.
  */
 function validateUnicodeGroupName(name, state) {
-  const isUnicodeName = uRe.test(name) || ucpRe.test(name);
+  const isUnicodeName = ucpReAnywhere.test(name);
   const isUnicodeState = (state === 'u' || state === 'xu' || state === 'u_class');
 
   if (isUnicodeName && !isUnicodeState) {
     throw new SyntaxError(`invalid group Unicode name "${name}", use \`u\` flag.`);
   }
+  
+  return name;
+}
+
+// Matches the following production: https://tc39.es/ecma262/#prod-RegExpUnicodeEscapeSequence
+//
+//  RegExpUnicodeEscapeSequence ::
+//    `u` LeadSurrogate `\u` TrailSurrogate   # as 'leadSurrogate', 'trailSurrogate'
+//    `u` LeadSurrogate                       # as 'leadSurrogateOnly'
+//    `u` TrailSurrogate                      # as 'trailSurrogateOnly'
+//    `u` NonSurrogate                        # as 'nonSurrogate'
+//    `u` `{` CodePoint `}`                   # as 'codePoint'
+//
+//  LeadSurrogate ::
+//    Hex4Digits but only if the SV of Hex4Digits is in the inclusive range 0xD800 to 0xDBFF        # [dD][89aAbB][0-9a-fA-F]{2}
+//
+//  TrailSurrogate ::
+//    Hex4Digits but only if the SV of Hex4Digits is in the inclusive range 0xDC00 to 0xDFFF        # [dD][c-fC-F][0-9a-fA-F]{2}
+//
+//  NonSurrogate ::
+//    Hex4Digits but only if the SV of Hex4Digits is not in the inclusive range 0xD800 to 0xDFFF    # [0-9a-ce-fA-CE-F][0-9a-fA-F]{3}|[dD][0-7][0-9a-fA-F]{2}
+//
+//  CodePoint ::
+//    HexDigits but only if MV of HexDigits ≤ 0x10FFFF                                              # 0*(?:[0-9a-fA-F]{1,5}|10[0-9a-fA-F]{4})
+//
+const uidRe = /\\u(?:([dD][89aAbB][0-9a-fA-F]{2})\\u([dD][c-fC-F][0-9a-fA-F]{2})|([dD][89aAbB][0-9a-fA-F]{2})|([dD][c-fC-F][0-9a-fA-F]{2})|([0-9a-ce-fA-CE-F][0-9a-fA-F]{3}|[dD][0-7][0-9a-fA-F]{2})|\{(0*(?:[0-9a-fA-F]{1,5}|10[0-9a-fA-F]{4}))\})/;
+
+function decodeUnicodeGroupName(name) {
+  return name.replace(new RegExp(uidRe, 'g'), function (_, leadSurrogate, trailSurrogate, leadSurrogateOnly, trailSurrogateOnly, nonSurrogate, codePoint) {
```

Comment:
By adding a `/g` flag, the `lastIndex` is preserved between calls, even in the event of an unhandled exception. Rather than constantly mutating a value that is shared between multiple disparate calls, this ensures a fresh `RegExp` instance is created so that the `lastIndex` mutations remain local to the function.

Reply:
Gotcha, OK.